tmt/steps/discover/fmf.py

```diff
@@ -306,8 +311,13 @@ def assert_git_url(plan_name: Optional[str] = None) -> None:
             if dist_git_source:
                 # Ensure we're in a git repo when extracting dist-git sources
                 try:
+                    assert self.step.plan.my_run is not None  # narrow type
+                    assert self.step.plan.my_run.tree is not None  # narrow type
+                    assert self.step.plan.my_run.tree.root is not None  # narrow type
                     git_root = get_git_root(self.step.plan.my_run.tree.root)
                 except tmt.utils.RunError:
+                    assert self.step.plan.my_run is not None  # narrow type
+                    assert self.step.plan.my_run.tree is not None  # narrow type
```

Comment:
I believe that's a valid observation and concern. I believe there is a solution, although hard to pick one.

`Plan.my_run` is `Optional[Run]`, because `Plan.__init__()` accepts `run: Optional[Run]`. One way would be to review each and every place where `Plan` is instantiated and find out whether `run` is even ever `None` - maybe it's always defined.

If it may be `None`, it'd be helpful to identify those situations, and we might split `my_run` into two attributes: one for regular use, which would apply `assert` internally and always return `Run`, and one for those special occasions, keeping the original `Optional[Run]` return type.

Or a subclass for those special cases, with a much narrower `my_run: None` (and changing the original to `Plan.my_run: Run`), to use them in different code flows. That would save us from having two different attributes I suggested above.

This requires a deeper knowledge of tmt internals than I have, and a dedicated review of code paths, and it's the reason why there's a "review all `assert`s" step in the annotations issue, so we could plan the next steps to reduce this very complexity.

Reply:
One more thing I just ran into: adding annotations to fmf library could also help clear the situation. One of the repeating `assert`s is to narrow `fmf.Tree.root` to `str` - according to analysis, it's `Optional[str]` (it even starts as `self.root = None`).

Reply:
Good to hear there are some options. Sounds like the extra subclass `PlanWithoutRun` with an explicit `my_run: None` could be a solution. Let's keep that for another PR as well.